### PR TITLE
Bug 1880926: Stop rescheduling non-existing nodes

### PR DESCRIPTION
**Pull request by periklis**
### Description
This PR is addressing the case where deployments/statefulsets for Elasticsearch nodes are not existing although the last recorded status registered them as unschedulable. This event may happen during cluster upgrades where OCP cluster nodes become unavailable and in turn ES cluster nodes unschedulable for a period of time. During a cluster upgrade deployments/statefulset may get deleted and recreated because their owner references get deleted and recreated.

In more detail the implementation takes into account that deployments/statefulsets may not exist anymore and stops addressing them as unschedulable and re-creates anew. The rationale to re-create the nodes immediately in `progressUnschedulableNodes` is to bypass the need to actively change the node status from `Unschedulable` to something like `ScheduledForRedeploy` to get re-creates. This sounds like a code smell misusing the redeploy routines.

/cc @blockloop @ewolinetz 
/assign @ewolinetz 
/cherry-pick release-4.5
 
#### Links
- Bugzilla:  https://bugzilla.redhat.com/show_bug.cgi?id=1880926

### CODE REVIEW

File: pkg/k8shandler/cluster.go [hunk 1]
```diff
@@ -200,17 +204,36 @@ func (er *ElasticsearchRequest) getNodeUpgradeInProgress() NodeTypeInterface {
...
+			for _, node := range clusterNodes {
+				if nodeStatus.DeploymentName == node.name() || nodeStatus.StatefulSetName == node.name() {
+
+					var obj runtime.Object
+					switch node.kind() {
+					case "Deployment":
+						obj = &appsv1.Deployment{}
+					case "StatefulSet":
+						obj = &appsv1.StatefulSet{}
+					}
```
Comment (periklis) [inline pkg/k8shandler/cluster.go:223]: Need to hide this behind the implementation. It is currently brittle.

Comment (periklis) [reply pkg/k8shandler/cluster.go]: This is resolved now by `node.isMissing()`

File: pkg/k8shandler/cluster.go [hunk 2]
```diff
@@ -200,17 +201,22 @@ func (er *ElasticsearchRequest) getNodeUpgradeInProgress() NodeTypeInterface {
...
+	for _, nodeStatus := range cluster.Status.Nodes {
+		if isPodUnschedulableConditionTrue(nodeStatus.Conditions) ||
+			isPodImagePullBackOff(nodeStatus.Conditions) ||
+			isPodCrashLoopBackOff(nodeStatus.Conditions) {
+
+			for _, node := range clusterNodes {
+				if nodeStatus.DeploymentName == node.name() || nodeStatus.StatefulSetName == node.name() {
+					if node.isMissing() {
+						log.Info("Unschedulable node not existing, creating node", "node", node.name())
+						return node.create()
```
Comment (ewolinetz) [inline pkg/k8shandler/cluster.go:215]: we don't want to create here, we want to just skip trying to progress it

Comment (periklis) [reply pkg/k8shandler/cluster.go:215]: As mentioned on voice and in the description above, the routines after `progressUnshedulableNodes` require conditions with `UpgradeStatus: CertRedploy` and from my smoke tests so far it is faster to recover a lost deployment here.

Comment (ewolinetz) [reply pkg/k8shandler/cluster.go:215]: i'm not following your comment about requiring conditions...
having the create in a place where it is seemingly hidden (its not obvious that we would create here) concerns me

Comment (periklis) [reply pkg/k8shandler/cluster.go]: Based on prior off-the-record discussion, I move the back to skipping progressing upgrade nodes if the node is not existing. In addition the status update routines will prune `status.nodes` and `status.pods` for non-existing nodes first and then update the fields for existing ones.

File: pkg/k8shandler/cluster.go [hunk 3]
```diff
@@ -200,17 +201,22 @@ func (er *ElasticsearchRequest) getNodeUpgradeInProgress() NodeTypeInterface {
...
+
+	for _, nodeStatus := range cluster.Status.Nodes {
+		if isPodUnschedulableConditionTrue(nodeStatus.Conditions) ||
+			isPodImagePullBackOff(nodeStatus.Conditions) ||
+			isPodCrashLoopBackOff(nodeStatus.Conditions) {
+
+			for _, node := range clusterNodes {
+				if nodeStatus.DeploymentName == node.name() || nodeStatus.StatefulSetName == node.name() {
+					if node.isMissing() {
+						log.Info("Unschedulable node not existing, skip progressing node changes", "node", node.name())
```
Comment (ewolinetz) [inline pkg/k8shandler/cluster.go:214]: nit but maybe reword to be something like "Unschedulable node does not have a k8s object, skipping"  ?

File: pkg/k8shandler/status.go [hunk 4]
```diff
@@ -261,147 +264,183 @@ func (er *ElasticsearchRequest) updateNodeConditions(status *api.ElasticsearchSt
...
 			}
+		}
+	}
+}
 
-			if !thresholdEnabled {
-				// disk threshold is not enabled, continue to next node
-				continue
-			}
+func (er *ElasticsearchRequest) updatePodStateMap(nodeName string, nodeStatus *api.ElasticsearchNodeStatus, pods []v1.Pod, thresholdEnabled bool) {
```
Comment (ewolinetz) [inline pkg/k8shandler/status.go:325]: this function name is a little misleading since it does more than just update the pod state map, it also updates the node conditions

File: pkg/k8shandler/cluster.go (hunk 3 above)
Comment (periklis) [reply pkg/k8shandler/cluster.go]: sure!

File: pkg/k8shandler/status.go (hunk 4 above)
Comment (periklis) [reply pkg/k8shandler/status.go]: Renamed to `updatePodNodeConditions`. WDYT?

File: pkg/k8shandler/status.go [hunk 5]
```diff
@@ -261,147 +264,183 @@ func (er *ElasticsearchRequest) updateNodeConditions(status *api.ElasticsearchSt
...
+	return nil
+}
+
+func (er *ElasticsearchRequest) pruneStatus(status *api.ElasticsearchStatus, nodeIndex int, nodeName string, pods []v1.Pod) {
+	clusterNodes := nodes[nodeMapKey(er.cluster.GetName(), er.cluster.GetNamespace())]
+	nodeStatus := status.Nodes[nodeIndex]
+	ll := er.L().WithValues("node", nodeStatus)
+
+	for _, node := range clusterNodes {
+		if nodeName == node.name() && node.isMissing() {
```
Comment (blockloop) [inline pkg/k8shandler/status.go:304]: The nesting here is too deep. Please invert the if condition here and on line 309 and use e.g.`if nodeName != node.name() || node.isMissing() { continue }`

**Review by jcantrill — COMMENTED** [review]
/approve
